### PR TITLE
GEODE-9362: Implement zadd with INCR option.

### PR DESCRIPTION
**Pull request by pivotal-eshu**
* allow zadd to return byte[] if INCR option is specified.
  * fixes a race issue of ZAddExecutorState in ZAddExecutor
  * added an ignored dunit test case showing score incrmented multiple times due to retry

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    doZAddIncrForAllMembers(increment1, increment1);
+
+    stopNodeWithPrimaryBucketOfTheKey(false);
+
+    doZCardWithRetries();
+    doZAddIncrForAllMembers(increment2, total);
+    verifyZScores();
+  }
+
+  private void doZAddIncrForAllMembers(double increment1, double increment12) {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java:126]: Typo here, I think it should probably be `increment2` not `increment12`.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java [hunk 2]
```diff
@@ -301,6 +306,106 @@ public void shouldUpdateScore_whenSettingMemberThatAlreadyExists() {
...
+  public void zaddIncrOptionDoseNotAddANewMemberWithXXOption() {
+    Object result = jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, "XX",
+        incrOption, "1", member);
+    assertThat(result).isNull();
+    assertThat(jedis.zscore(SORTED_SET_KEY, member)).isNull();
+  }
+
+  private final double initial = 355.681000005;
+  private final double increment = 9554257.921450001;
+  private final double expected = initial + increment;
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:333]: Could these fields be moved to the top of the class for consistency with other classes please?

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java [hunk 3]
```diff
@@ -46,11 +45,17 @@ public RedisResponse executeCommand(Command command, ExecutionHandlerContext con
...
 
-    long retVal = redisSortedSetCommands.zadd(command.getKey(),
+    Object retVal = redisSortedSetCommands.zadd(command.getKey(),
         new ArrayList<>(commandElements.subList(optionsFoundCount + 2, commandElements.size())),
         makeOptions(zAddExecutorState));
+    if (zAddExecutorState.incrFound) {
+      if (retVal == null) {
+        return RedisResponse.nil();
+      }
+      return RedisResponse.string((byte[]) retVal);
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java:55]: The [documentation for ZADD with INCR option](https://redis.io/commands/ZADD#return-value) states that the response will be a Bulk string response, not a simple string. I think this should be therefore be `RedisResponse.bulkString(retVal);` The documentation for ZINCRBY also says the same, so I think we may need to change the `RedisResponse` type in the `ZIncrByExecutor` as well, since it's also currently just using a simple string response.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+  }
+
+  private void doZCardWithRetries() {
+    int maxRetryAttempts = 10;
+    int retryAttempts = 0;
+    while (!zCardWithRetries(retryAttempts, maxRetryAttempts)) {
+      retryAttempts++;
+    }
+  }
```
Comment (sabbey37) [inline geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java:138]: It might be a bit more straightforward to take out the extra `zCardWithRetries` method and just do the following:
```
  private void doZCardWithRetries() {
    int maxRetryAttempts = 10;
    int retryAttempts = 0;

    while (retryAttempts < maxRetryAttempts) {
      long memberSize;
      try {
        memberSize = jedis.zcard(sortedSetKey);
      } catch (JedisClusterMaxAttemptsException e) {
          retryAttempts++;
          if(retryAttempts == maxRetryAttempts) {
            throw e;
          }
          continue;
      }
      assertThat(memberSize).isEqualTo(setSize);
      break;
    }
  }
 ```
 ...but that might just be a personal preference.

**Review by sabbey37 — COMMENTED** [review]
Looks great, Eric!  Just a few suggestions.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java (hunk 3 above)
Comment (sabbey37) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java:55]: Good catch!  I wonder if there are other places where we're returning simple string instead of bulk string...

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java [hunk 5]
```diff
@@ -35,12 +35,17 @@
...
 import org.apache.geode.redis.internal.RedisConstants;
+import org.apache.geode.redis.internal.netty.Coder;
 import org.apache.geode.test.awaitility.GeodeAwaitility;
 
 public abstract class AbstractZAddIntegrationTest implements RedisIntegrationTest {
   private JedisCluster jedis;
+  private final String member = "member";
+  private final String incrOption = "INCR";
+
   private static final int REDIS_CLIENT_TIMEOUT =
```
[the remark this answers is not in the document]
Comment (sabbey37) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:49]: Oooh we also have the `REDIS_CLIENT_TIMEOUT` variable defined in `RedisClusterStartupRule` which is in the `commonTest` module... so we could potentially remove it from `RedisIntegrationTest` altogether.

On line 57 below this we use "localhost" when starting the JedisCluster, we could import the `BIND_ADDRESS` variable from `RedisClusterStartupRule` and use that instead.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java (hunk 3 above)
Comment (DonalEvans) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java:55]: Do we know if there are any (possibly currently disabled) tests in the `redis-test-openjdk11` precheckin job that would fail due to wrong return format?

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java [hunk 6]
```diff
@@ -35,12 +35,17 @@
...
 import redis.clients.jedis.params.ZAddParams;
 
+import org.apache.geode.redis.ConcurrentLoopingThreads;
 import org.apache.geode.redis.RedisIntegrationTest;
 import org.apache.geode.redis.internal.RedisConstants;
+import org.apache.geode.redis.internal.netty.Coder;
 import org.apache.geode.test.awaitility.GeodeAwaitility;
 
 public abstract class AbstractZAddIntegrationTest implements RedisIntegrationTest {
   private JedisCluster jedis;
```
Comment (nonbinaryprogrammer) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:45]: I'd like to see a test for when increment is infinity, and when incrementing a score of infinity, since those are both valid. It would need to test negative and positive infinity, and that you can pass in "Infinity", "+Infinity", "-Infinity", "infinity", "+infinity", "-infinity", "inf", "+inf", and "-inf" as increments.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java [hunk 7]
```diff
@@ -301,6 +306,106 @@ public void shouldUpdateScore_whenSettingMemberThatAlreadyExists() {
...
+        jedis.sendCommand(SORTED_SET_KEY, Protocol.Command.ZADD, SORTED_SET_KEY, incrOption,
+            Coder.doubleToString(increment), member);
+
+    assertThat(Coder.bytesToDouble((byte[]) result)).isEqualTo(expected);
+    result = jedis.zscore(SORTED_SET_KEY, member);
+    assertThat(result).isEqualTo(expected);
+  }
+
+  @Test
+  public void zaddIncrOptionCanIncrementScoreConcurrently() {
```
Comment (nonbinaryprogrammer) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:387]: might be a good idea to add a test for concurrent modification of the same key over and over again, making sure that all the increments get applied correctly. I would do that with a random increment every time, then storing that to an atomic integer.

**Review by nonbinaryprogrammer — CHANGES_REQUESTED** [review]

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java (hunk 3 above)
Comment (sabbey37) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZAddExecutor.java:55]: I'm not sure that there are.  I ran a few of the zset.tcl tests against the current implementation (`test "ZINCRBY return value"`, `test "ZINCRBY - increment and decrement`, and `test "ZINCRBY - can create a new sorted set`.. and a few others), but they all passed.

The difference is obvious on the CLI, here's a bulk string returned from native Redis:
```
127.0.0.1:6379> zadd a incr ch 1 a
"3"
```

Here's a simple string returned by us:
```
127.0.0.1:6380> zadd a incr 1 a
1
```
We could write tests at a lower level, but I'm not sure if it's super valuable given that clients are able to handle either one in the same fashion.

File: geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      if (retries < maxRetries) {
+        return false;
+      }
+      throw e;
+    }
+    assertThat(memberSize).isEqualTo(setSize);
+    return true;
+  }
+
+  private boolean hitJedisClusterIssue2347 = false;
```
[the remark this answers is not in the document]
Comment (pivotal-eshu) [reply geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java:154]: The boolean value hitJedisClusterIssue2347 needs to be referenced by to the caller. There does not seems to be a clean way to do it.

File: geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    verifyZScores();
+  }
+
+  private void verifyZScores() {
+    for (int i = 0; i < setSize; i++) {
+      assertThat(jedis.zscore(sortedSetKey, baseMemberName + i)).isEqualTo(total);
+    }
+  }
+
+  private void doZAddIncr(int i, double increment, double total, boolean isConcurrentExecution) {
```
[the remark this answers is not in the document]
Comment (pivotal-eshu) [reply geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java:104]: It is being called in doZAddIncrForAllMembers(increment1, increment1);

File: geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sabbey37) [reply geode-apis-compatible-with-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/sortedset/ZAddIncrOptionDUnitTest.java:154]: You could do it like this:

```
  public void zAddWithIncrOptionCanIncrementScoresDuringPrimaryIsCrashed() throws Exception {
    AtomicBoolean hitJedisClusterIssue2347 = new AtomicBoolean(false);
    doZAddIncrForAllMembers(increment1, increment1);

    Future<Void> future1 = executor.submit(() -> doZAddIncrForAllMembersDuringCrash(hitJedisClusterIssue2347));
    Future<Void> future2 = executor.submit(() -> stopNodeWithPrimaryBucketOfTheKey(true));

    future1.get();
    future2.get();

    if (!hitJedisClusterIssue2347.get()) {
      doZCardWithRetries();
      verifyZScores();
    }
  }

  private void doZAddIncrForAllMembersDuringCrash(AtomicBoolean hitJedisClusterIssue2347) {
    for (int i = 0; i < setSize; i++) {
      try {
        doZAddIncr(i, increment2, total, false);
      } catch (JedisClusterMaxAttemptsException ignore) {
        hitJedisClusterIssue2347.set(true);
      }
    }
  }
```

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java (hunk 6 above)
Comment (pivotal-eshu) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:45]: As zadd with incr option internally calls the zincrby methods. I think the test coverage for zincrby should be good enough already.

**Review by DonalEvans — APPROVED** [review]

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java (hunk 6 above)
Comment (nonbinaryprogrammer) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZAddIntegrationTest.java:50]: good point 👍

**Review by nonbinaryprogrammer — APPROVED** [review]
looks good

**Review by sabbey37 — APPROVED** [review]
